src/clients/dotnet/TigerBeetle.Tests/UInt128Tests.cs

```diff
@@ -109,14 +106,40 @@ public void BigIntegerExceedU128()
     [TestMethod]
     public void LittleEndian()
     {
-        var expected = new byte[16] { 86, 52, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+        var expected = new byte[16] {86, 52, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
 
         Assert.IsTrue(expected.SequenceEqual(expected.ToUInt128().ToArray()));
         Assert.IsTrue(expected.SequenceEqual(BigInteger.Parse("123456", NumberStyles.HexNumber).ToUInt128().ToArray()));
-        Assert.IsTrue(expected.SequenceEqual(new Guid(expected).ToUInt128().ToArray()));
+        // `bigEndian: true` signals to the Guid constructor to store the bytes as-is 
+        Assert.IsTrue(expected.SequenceEqual(new Guid(expected, bigEndian: true).ToUInt128().ToArray()));
         Assert.IsTrue(expected.SequenceEqual(new UInt128(0, 0x123456).ToArray()));
     }
 
+    [TestMethod]
+    public unsafe void EndToEndTigerBeetleSimulationViaGuid()
+    {
+        // "00000001-0001-4000-AA00-000000000000" 
+        var idSourcedExternally = Guid.Parse("00000001-0001-4000-AA00-000000000000"); // e.g. guid from other database
+        var bytesStoredInTB = new byte[]
+            {0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x40, 0x00, 0xAA, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+
+        // Parse TB result (TB client reads by overlaying dotnet types on raw bytes)
+        UInt128 parsed = UInt128.Zero;
+        bytesStoredInTB.CopyTo(new Span<byte>(&parsed, 16));
+
+        // Round-trip via GUID
+        var roundTripped = parsed.ToGuid().ToUInt128();
+
+        // Send bytes to TB (TB client send raw bytes of dotnet types over the wire) 
+        var roundTrippedBytes = new Span<byte>(&roundTripped, 16);
+
+        var externalIdUInt128 = idSourcedExternally.ToUInt128();
+        var externalIdBytes = new Span<byte>(&externalIdUInt128, 16);
+
+        Assert.IsTrue(roundTrippedBytes.SequenceEqual(bytesStoredInTB));
+        Assert.IsTrue(externalIdBytes.SequenceEqual(bytesStoredInTB));
```

Comment:
I'm sorry, but I'm still failing to see how the boundary between the bytes and the string representation of a Guid would be mixed in a "round trip".

Considering this test below (passing in the current implementation), the correct behavior seems to be `bigEndian = false`, the same logic for `BigInteger` as well.

Probably I am missing something here :thinking: 

```c#
    [TestMethod]
    public void TestEndianness()
        // Little-endian bytes of `@as(u128, 0x123456)`:
        var expectedBytes = new byte[16] { 86, 52, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };

        // Guid string representation of the same binary value:
        var expectedGuid = Guid.Parse("00123456-0000-0000-0000-000000000000");

        // Literal representation in hex:
        UInt128 expectedU128 = 0x123456;

        // BigInt representation.
        var expectedBigInt = BigInteger.Parse("123456", NumberStyles.HexNumber);

        // Asserting our .Net understanding of Guids:
        // "bigEndian: false" doesn't change the bytes
        bool le = expectedBytes.SequenceEqual(new Guid(expectedBytes, bigEndian: false).ToByteArray());
        Assert.IsTrue(le);
        // "bigEndian: true" changes the bytes
        bool be = expectedBytes.SequenceEqual(new Guid(expectedBytes, bigEndian: true).ToByteArray());
        Assert.IsFalse(be);


        // Asserting our .Net understanding of BigIntegers:
        //  BigInteger.ToByteArray returns the fewest number of bytes possible.

        // "isBigEndian: false" doesn't change the bytes
        var bigIntFewestBytesLE = expectedBigInt.ToByteArray(isUnsigned: true, isBigEndian: false);
        Assert.IsTrue(expectedBytes.Take(bigIntFewestBytesLE.Length).SequenceEqual(bigIntFewestBytesLE));
        Assert.IsTrue(expectedBytes.Skip(bigIntFewestBytesLE.Length).All(b => b == 0));
        // "isBigEndian: true" changes the bytes
        var bigIntFewestBytesBE = expectedBigInt.ToByteArray(isUnsigned: true, isBigEndian: true);
        Assert.IsFalse(expectedBytes.Take(bigIntFewestBytesBE.Length).SequenceEqual(bigIntFewestBytesBE));

        // Binary representation
        Assert.IsTrue(expectedBytes.SequenceEqual(expectedGuid.ToByteArray()));
        Assert.IsTrue(expectedBytes.SequenceEqual(expectedU128.ToArray()));

        // U128 representation
        Assert.IsTrue(expectedU128 == 0x123456UL);
        Assert.IsTrue(expectedU128 == expectedBytes.ToUInt128());
        Assert.IsTrue(expectedU128 == expectedGuid.ToUInt128());
        Assert.IsTrue(expectedU128 == expectedBigInt.ToUInt128());
        Assert.IsTrue(expectedU128 == (ulong)expectedBigInt);

        // Guid representation
        Assert.IsTrue(expectedU128.ToGuid() == expectedGuid);
        Assert.IsTrue(expectedU128 == expectedGuid.ToUInt128());
        Assert.IsTrue(expectedU128.ToGuid().ToString() == expectedGuid.ToString());

        // BigInt representation
        Assert.IsTrue(expectedU128 == expectedBigInt.ToUInt128());
        Assert.IsTrue(expectedU128.ToBigInteger() == expectedBigInt);
        Assert.IsTrue(expectedBytes.SequenceEqual(expectedBigInt.ToUInt128().ToArray()));
    }
```

Reply:
This is the crux of the issue
> How exactly the guid 00000001-0001-4000-AA00-000000000000 from another database would be stored in TigerBeetle as {0x00, 0x00, 0x00, 0x01, 0x00, 0x01, 0x40, 0x00, 0xAA, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, since all conversion methods represent it as { 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x40, 0xAA, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 } instead?

You're correct - Tigerbeetle client reads/writes the bytes as they are laid out in the guid type, but everything else in dotnet doesn't. See [mysql](https://github.com/mysql-net/MySqlConnector/blob/5a377426aa049f90689e3fffa9894ecd7e2f8e7b/src/MySqlConnector/ColumnReaders/GuidBinary16ColumnReader.cs) and [postgres](https://github.com/npgsql/npgsql/blob/main/src/Npgsql/Internal/Converters/Primitive/GuidUuidConverter.cs)

The reason they swap the bytes around is that an "abstract" guid/uuid value of say  `00000001-0001-4000-AA00-000000000000`  is laid out internally as `{ 0x01, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x40, 0xAA, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 }` in the ms guid type. The problem is that the rest of the world outside of microsoft  understands that series of bytes to represent a uuid value of `01000000-0100-0040-AA00-000000000000` and NOT `00000001-0001-4000-AA00-000000000000`.

Looking at your test:

<blockquote>

```
var expectedBytes = new byte[16] { 86, 52, 18, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };

// Guid string representation of the same binary value:
var expectedGuid = Guid.Parse("00123456-0000-0000-0000-000000000000");
```

</blockquote>

Dotnet would say that the string UUID corresponding to those bytes is `00123456-0000-0000-0000-000000000000`, so your test passes, and those bytes are written to TB. But javascript and java would say it's `56341200-0000-0000-0000-000000000000`.

I ran into the issue with the following steps -
- Generate a random guid to use as account ID
- Save account metadata to postgres
- Save account to TB
- Read GUID from postgres as uint128
- Query TB using the cli for account with that uint128 -> this fails with a NotFound error
- Query TB using the guid in postgres having swapped the bytes -> succeeds

..because the same guid  bytes were written to postgres in the uuid big-endian format, and written to tb in the ms guid format.

Reply:
My cli is now saying `message queue for peer message_bus.MessageBusType(.client).Connection.Peer{ .replica = 0 } full, dropping request message` but I was going to test adding an account with ID=123456 via cli and then querying using these 2 guids ->

- new Guid("")

Reply:
Sorry for going back and forth with this, but I'm trying to determine where this boundary is crossed and how transparently or explicitly we should handle this transition.

> ..because the same guid bytes were written to postgres in the uuid big-endian format, and written to tb in the ms guid format.

Let me check my understanding:

When generating the GUID "00123456-0000-0000-0000-000000000000" in .Net and saving it to PG/MySQL, one could read it back as "00123456-0000-0000-0000-000000000000" in C#, but "56341200-0000-0000-0000-000000000000" in Java or JS, right? Nevertheless, only the string representation diverges, the binary representation should still be consistent across different platforms, is it correct?

When you say "Read GUID from Postgres as uint128", is that the database performing the cast?
I mean, since it was a GUID originally, shouldn't you need to read it as a GUID and then call `.ToUInt128()` in C# to perform the proper conversion*?

_*Thanks for spotting the wrong spelling in conver**t**ion :smile:_

Reply:
Something to add, from windows-rs

https://github.com/microsoft/windows-rs/blob/15947886be041bebd0fe670fd4e73f18d95100d0/crates/libs/core/src/guid.rs#L50-L65

Reply:
> I was going to test adding an account with ID=123456 via cli and then querying using these 2 guids ->
> 
> * Guid.Parse("00123456-0000-0000-0000-000000000000")
> * Guid.Parse("56341200-0000-0000-0000-000000000000")
> 
> Which one should hit?

In C#, it should hit `Guid.Parse("00123456-0000-0000-0000-000000000000")`, but I think the string representation is the less important thing here (since it is an app abstraction to deal with `uint128` though a `[16]u8`). The important thing is that conversions must be coherent with the binary representation in all platforms (like the hex literal "0x123456" is converted to the same BigInt in C#, GO, Java, and JS).

Speaking of hex, you must create the account with ID=1193046, which is the decimal representation of 0x123456.

For the record, I'm convinced that we should fix it at the C# client, since a SELECT into PG would render this UUID as a different string ... It's a huge footgun.

Just need to setup an environment to test all the variants first, in order to avoid introducing another footgun elsewhere!

> I'll take a look next week

Likewise, thanks!

Reply:
> In C#, it should hit Guid.Parse("00123456-0000-0000-0000-000000000000"), but I think the string representation is the less important thing here (since it is an app abstraction to deal with uint128 though a [16]u8). The important thing is that conversions must be coherent with the binary representation in all platforms (like the hex literal "0x123456" is converted to the same BigInt in C#, GO, Java, and JS).

The binary representation is not consistent from dotnet to other languages. I'd generally agree that the string representation isn't important, but it happens to also be the binary representation in other languages, so it's a nice way to quickly verify

> When generating the GUID "00123456-0000-0000-0000-000000000000" in .Net and saving it to PG/MySQL, one could read it back as "00123456-0000-0000-0000-000000000000" in C#, but "56341200-0000-0000-0000-000000000000" in Java or JS, right? Nevertheless, only the string representation diverges, the binary representation should still be consistent across different platforms, is it correct?
 
`Guid.Parse("00123456-0000-0000-0000-000000000000")` will stored in PG as `00 12 34 56 00...`, but stored in TB as `56 34 12 00...`.

> When you say "Read GUID from Postgres as uint128", is that the database performing the cast?

I mean construct a little-endian UInt128 "by hand" (really I used a c# repl to run `BinaryPrimitives.ReadUInt128LittleEndian(Guid.Parse("uuid-string-copy-pasted-from-db").ToArray(bigEndian:true))` but you can use your language of choice here to construct a little-endian UInt128 from 128 bits).

> I mean, since it was a GUID originally, shouldn't you need to read it as a GUID and then call .ToUInt128() in C# to perform the proper conversion*?

The key here was to avoid using dotnet. I have 128 bits in PG, and I want to use those bits to query tb from another language for what the dotnet app had written earlier. My understanding is that inside of tigerbeetle, at the end of the day an ID is just 128 bits, what they mean or how they are handled (guid/uint128/byte array) exactly is up to the developer. The cli handles the 128 bits as a little endian UInt128 so that's the "other language" that I was using.